### PR TITLE
Combo hit

### PR DESCRIPTION
**Pull request by bechenard**
July 4th gift for Dave

### CODE REVIEW

**Review by brownd1978 — CHANGES_REQUESTED** [review]
Hi Bertrand, I guess the point of this PR is replacing the boost accumulators, not just re-indenting?  Sorry to be a nag, but the rule is that code without validated fhicl needs to migrated when it is changed.

**Comment by bechenard** [comment]
Hi Dave,

What is the coefficient of elasticity of this rule? I did replace the 
accumulators and changed some of the logic to make it easier to maintain.

Cheers,

Bertrand

**Comment by bechenard** [comment]
I fixed the fhicl validation, and hopefully all instances in the prologs. My limited tests worked, but let me know if there is an issue.

**Comment by brownd1978** [comment]
@FNALbuild run build tests

File: TrkHitReco/src/CombineStrawHits_module.cc [hunk 1]
```diff
@@ -1,226 +1,212 @@
...
+        throw cet::exception("RECO")<<"mu2e::CombineStrawHits: No ComboHit collection found for tag" <<  _chTag << std::endl;
+      const ComboHitCollection* chcolOrig = chH.product();
+
+      auto chcolNew = std::make_unique<ComboHitCollection>();
+      chcolNew->reserve(chcolOrig->size());
+      chcolNew->setParent(chH);
+
+      combine(chcolOrig, *chcolNew);
+std::cout<<"CH "<<chcolNew->size()<<std::endl;
+for (auto& ch : *chcolNew) std::cout<<ch.time()<<" "<<ch.pos()<<std::endl;
```
Comment (bonventre) [inline TrkHitReco/src/CombineStrawHits_module.cc:108]: Looks like some leftover debugging print statements

**Comment by bechenard** [comment]
Crap, I was certain to have removed them... I'll fix this later.

Bertrand

**Comment by kutschke** [comment]
Bertrand:  The conflict is generated by the "Offline shift" and will be straightforward to fix: just add Offline/ as a prefix to all includes in CombineStrawHits_module.cc .  Probably the easiest thing is to merge/rebase.  If this PR is otherwise really complete we can fix it on the GitHub website with the online editor - but I don't know what happens if we fix it there and you decide you need another commit on this PR.  Do you want to use this as a chance to learn what happens?

**Comment by bechenard** [comment]
Hi Rob,

I fixed ComboHits, but FastCaloCluster will require more work (I'll 
cancel the PR and remake one when it's ready). I'm away until the end of 
July and will have limited time to work on this.

Cheers,

Bertrand

**Comment by kutschke** [comment]
@FNALbuild run build tests

**Comment by kutschke** [comment]
This CI build failed because the code still has one old style include statement.  See:

https://github.com/bechenard/Offline/blob/61db19f3afd0011641ecad71686039172d7da32a/TrkHitReco/src/CombineStrawHits_module.cc#L13

The same file is already properly included 4 lines earlier.  Just delete this line, commit and push.  Then I will relaunch the CI.

**Comment by bechenard** [comment]
Either very late tonight or this weekend. Feel free to do it if you can

**Comment by kutschke** [comment]
I don't know how to fix this one from the GitHub end.  I can only do that in a subset of the cases that GitHub detects a conflict.  In this case GitHub gives it the green light but our CI fails during the build.  So whenever you get to it.

**Comment by bechenard** [comment]
Ok, made the fix, that should do it.

**Comment by brownd1978** [comment]
@FNALbuild run build tests

**Comment by bechenard** [comment]
Ok, now it is really gone...

Bertrand

**Comment by kutschke** [comment]
@FNALbuild run build tests

**Review by brownd1978 — APPROVED** [review]

File: TrkHitReco/src/CombineStrawHits_module.cc [hunk 2]
```diff
@@ -1,226 +1,209 @@
...
+          combohit.init(hit1,ich);
+          int panel1 = hit1.strawId().uniquePanel();
+
+          for (size_t jch=ich+1;jch<chcolOrig->size();++jch)
+          {
+              if (isUsed[jch]) continue;
+              const ComboHit& hit2 = (*chcolOrig)[jch];
+
+              if (hit2.strawId().uniquePanel() != panel1) break;
+              if (hit2.strawId().straw()-hit1.strawId().straw()> _maxds ) break;
```
Comment (bonventre) [inline TrkHitReco/src/CombineStrawHits_module.cc:131]: Don't we still need an abs() here?

**Comment by bechenard** [comment]
Hi Richie,

According to Dave, there is a guarantee that the hits are ordered by 
panels/straws/... , so strawID of hit2 is always >= strawID of hit1. But 
I should triple check.

Cheers,

Bertrand

**Review by bonventre — APPROVED** [review]

**Pull request merged**